### ISSUE TITLE
shinyFiles not binding with renderUI output

### ISSUE THREAD

**Issue by kajomano** [thomasp85/shinyFiles#25]
Hi Thomas,

I tried to use your package in a dynamic shiny page rendering the output with renderUI, but the observer does not seem to trigger when clicking the button. The same code works if the shinyFilesButton() function is called in the ui somewhere.

``` r
shinyApp(
  ui = fluidPage(
    sidebarLayout(
      sidebarPanel(uiOutput("loadFileOutPut")),
      mainPanel()
    )
  ), 
  server = function(input, output, session) {
    shinyFileChoose( input = input, id = "loadFile", session = session, roots = c( "project" = getwd()))

    output$loadFileOutPut <- renderUI({
      shinyFilesButton( "loadFile", "Load a file", "Load a file", multiple = FALSE ) 
    })
  }
)
```

**Comment by kajomano** [comment]
This guy has the same issue: http://stackoverflow.com/questions/33740939/shinyfiles-and-renderui-dont-work-properly

**Comment by thomasp85** [comment]
This is definitely something I want to look into - thanks for bringing it to my attention! Currently, and until early summer, I’m forced to focus on finishing my PhD, so all my fun side projects (of which shinyFiles is one), has been severely downgraded in priority. If it is mission critical for you I’ll urge you to make a PR with a fix, and I’ll take the time to review it. My gut feeling is that it is related to the unbind/bindall pattern exposed by Shiny - It has caused problems on other occasions when they’ve changed implementation...

**Comment by kajomano** [comment]
Hey, thanks for the reply. While fixing the issue would make my life easier, it's really not mission critical. I am somewhat new to shiny and web developing in general, but if I have the time I will definitely try to do a PR.

**Comment by kajomano** [comment]
Hi Thomas,

I have been trying to make your package work with renderUI and I have found the following problems so far (only using FileChooser):

1) When using a renderUI call to create the shinyFilesButton on the page, the corresponding shinyFileChoose observer might be created before the HTML element is actually present on the page (populateFileChooser gets an empty element selection). This results in the browser not being populated initially, since that is done by the observer.

2) As you guessed, the Shiny.bindAll() call in createFileChooser() somehow removes all the elements binded data.

Manually populating the FileChooser after creation and backing up and rebinding the button's data after the bindAll() call makes the root directories' files show up, but the filebrowser is still unresponsive and after closing you can't reopen the browser.

**Comment by darkwish121** [comment]
Hi Thomas, This issue is still alive. Are you going to fix or hint me where to use Shiny.unbindAll() and Shiny.bindAll() in your code. I'll fork and do it. [Here](http://stackoverflow.com/questions/36197354/rendergvis-not-working-with-renderui-and-plotoutput-while-creating-multiple-go) is my question related to this in stackoverflow

**Comment by thomasp85** [comment]
I know -  It's in top three to-do after I hand in my PhD next month... I'm quite convinced that it is a regression in Shiny, rather than a bug in shinyFiles, but if you wish to try with a PR I'm very welcome to that

**Comment by darkwish121** [comment]
Thanks and good luck for your PHD

**Comment by dominiqueemmanuel** [comment]
Hi,

Do you have any information (from Shiny developers ?) if the problem came from a regression in Shiny finaly ?

Best regards,
Dom

**Comment by thomasp85** [comment]
No, but my best guess is that it happened with the 0.13 release. This release came with another regression related to bindAll so I believe there has been some rework done on the internals at that point. I've always made my UI with custom html so this has never been something I've bumped into myself, thus this is pure speculation

**Comment by demetriusgnu** [comment]
Hi, no solution for this? Thanks

**Comment by vnijs** [comment]
Made some progress here. But working with renderUI unfortunately still seems to be an issue. To install the PR with the updated code `shinyFiles` use:

`devtools::install_github("vnijs/shinyFiles")`

Initially, the modal linked to the `Save file` button will show files and folders. However, after changing the file type and clicking the `Save file` button a second time, no files or folders are shown.

```r
library(shiny)
library(shinyFiles)

# options(shiny.trace = TRUE)

shinyApp(
    ui = fluidPage(
        sidebarLayout(
            sidebarPanel(
                uiOutput("saveFileOutput"),
                selectInput("filetype", "Choose file type", c("csv", "txt", "rda"))
            ),
            mainPanel()
        )
    ),
    server = function(input, output, session) {
        volumes <- c("Project" = getwd())
        
        shinyFileSave(input, "saveFileUI", roots = volumes, session = session)
        
        output$saveFileOutput <- renderUI({
            shinySaveButton("saveFileUI", "Save file (renderUI)", "Save file as ...", filetype = input$filetype)
        })
    }
)
```

**Comment by thomasp85** [comment]
closed by 1c82b5e4c67565afc740cf925c9eb98962db7e6d
